### PR TITLE
Move wayland client to UWAC

### PR DESCRIPTION
**Pull request by hardening**
In this PR instead of reimplementing all wayland tricks, we rely on the UWAC library that do it for us.
With this PR the wayland client works better, it still remain an issue with keyboard handling (how to choose a RDP locale from the mmap file of wayland...).

### CODE REVIEW

**Comment by bmiklautz** [comment]
@hardening to build with uwac we need the lib somehow on the build systems otherwise the detection should disable wayland.

**Comment by hardening** [comment]
Ok so that means that my cmake detection is crappy. Working on this.

**Comment by bmiklautz** [comment]
@hardening the build works again but now we wouldn't do any test builds of the wayland client because uwac isn't available. - right?

**Comment by hardening** [comment]
@bmiklautz yeah right you caught me :)

**Comment by hardening** [comment]
@bmiklautz so this is the version with UWAC bundled in FreeRDP itself. Would be nice if xkbcommon could be installed on the builder...

**Comment by bmiklautz** [comment]
@hardening it's on the list I take care of it as soon as I can :).

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by bmiklautz** [comment]
@hardening would you rebase this to the latest master?

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by hardening** [comment]
@freerdp-bot I think you failed the last test, would you retry ?

File: cmake/FindWayland.cmake [hunk 1]
```diff
@@ -28,38 +25,35 @@
...
-          DOC "The Wayland client library"
+find_path(WAYLAND_INCLUDE_DIR wayland-client.h
+    HINTS ${WAYLAND_CLIENT_PKG_INCLUDE_DIRS}
 )
 
-include(FindPackageHandleStandardArgs)
-FIND_PACKAGE_HANDLE_STANDARD_ARGS(WAYLAND DEFAULT_MSG WAYLAND_LIBRARY WAYLAND_INCLUDE_DIR)
+find_library(WAYLAND_LIBS 
+    NAMES "wayland-client"
+    HINTS "${WAYLAND_CLIENT_PKG_LIBRARIES}"
```
Comment (akallabeth) [inline cmake/FindWayland.cmake]: @hardening According to cmake documentation this variable only contains the name of the library suitable to use with ```-l <libname>```, you should use ```WAYLAND_CLIENT_PKG_LIBRARY_DIRS``` instead

File: cmake/FindWayland.cmake [hunk 2]
```diff
@@ -28,38 +25,35 @@
...
+    HINTS ${XKBCOMMON_PC_INCLUDE_DIRS}
+)
 
-if(WAYLAND_FOUND)
-  set(WAYLAND_LIBRARIES ${WAYLAND_LIBRARY})
-  set(WAYLAND_INCLUDE_DIRS ${WAYLAND_INCLUDE_DIR})
-endif()
+find_library(XKBCOMMON_LIBS 
+    NAMES xkbcommon
+    HINTS "${XKBCOMMON_PC_LIBRARIES}"
```
Comment (akallabeth) [inline cmake/FindWayland.cmake]: @hardening Same issue as above.

**Comment by akallabeth** [comment]
@hardening This pull does not build properly on my debian jessie system. I have wayland libs installed, but they are not detected (see comments above).
Further the failed wayland detection does not disable all uwac related components (or abort the cmake run) and the compilation aborts with missing ```-luwac``` messages.

**Comment by hardening** [comment]
@akallabeth I have fixed the detection, but after deinstalling wayland libs I can't reproduce the compilation bug: for me the wayland client and libs are correctly skipped. Anyway my debian, VM is a wheezy not a jessie.

**Comment by hardening** [comment]
@freerdp-bot what's wrong with this PR, don't you want to test wayland ?

**Comment by bmiklautz** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@hardening can build the library on my machine now, the issues are (mostly) resolved, but you need to add ```${CMAKE_BINARY_DIR}/uwac/libuwac``` to the include path, as the protocol headers are generated there for out of tree builds.

File: uwac/libuwac/uwac-window.c [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+		return UWAC_ERROR_INTERNAL;
+	}
+
+	data = mmap(NULL, allocSize * nbuffers, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+	if (data == MAP_FAILED)	{
+		ret = UWAC_ERROR_NOMEMORY;
+		goto error_mmap;
+	}
+
+	pool = wl_shm_create_pool(w->display->shm, fd, allocSize * nbuffers);
```
Comment (bmiklautz) [inline uwac/libuwac/uwac-window.c]: Missing check?

File: uwac/libuwac/uwac-tools.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+#include <string.h>
+#include <uwac/uwac-tools.h>
+
+/** @brief */
+struct uwac_touch_automata {
+	struct wl_array tp;
+};
+
+void UwacTouchAutomataInit(UwacTouchAutomata *automata) {
+	wl_array_init(&automata->tp);
```
Comment (bmiklautz) [inline uwac/libuwac/uwac-tools.c]: Does this always succeed?

File: uwac/libuwac/uwac-input.c [hunk 5]
```diff
@@ -0,0 +1,845 @@
...
+}
+
+static void
+seat_handle_name(void *data, struct wl_seat *seat, const char *name)
+{
+	UwacSeat *input = data;
+	if (input->name)
+		free(input->name);
+
+	input->name = strdup(name);
```
Comment (bmiklautz) [inline uwac/libuwac/uwac-input.c]: The return value of name should be checked.

File: client/Wayland/wlfreerdp.c [hunk 6]
```diff
@@ -73,21 +90,23 @@ static BOOL wl_end_paint(rdpContext* context)
 	h = gdi->primary->hdc->hwnd->invalid->h;
 
 	context_w = (wlfContext*) context;
-	display = context_w->display;
-	window = context_w->window;
 
+	data = UwacWindowGetDrawingBuffer(context_w->window);
```
Comment (bmiklautz) [inline client/Wayland/wlfreerdp.c]: I'm not sure but shouldn't we add a check if data is set here?

**Comment by bmiklautz** [comment]
Looks good for me.
Except my comments +1.

File: uwac/libuwac/uwac-tools.c (hunk 4 above)
Comment (hardening) [inline uwac/libuwac/uwac-tools.c]: yes

**Comment by bmiklautz** [comment]
@hardening nice work! - I gonna add some versioning and pkg-config for it \o/

**Pull request merged**